qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,118 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Linear Function."""
+
+from typing import Union, List, Optional
+import numpy as np
+from qiskit.circuit import QuantumCircuit, Gate
+from qiskit.circuit.exceptions import CircuitError
+
+
+class LinearFunction(Gate):
+    """An n_qubit linear circuit."""
+
+    def __init__(
+        self,
+        linear: Union[List[List[int]], np.ndarray, QuantumCircuit],
+        validate_input: Optional[bool] = False,
+    ) -> None:
+        """Create a new linear function.
+
+        Internally, represents a linear function acting on n qubits as a n x n matrix of 0s and 1s.
+
+        Args:
+            linear (list[list] or ndarray or QuantumCircuit):
+                either an n x n matrix, describing the linear function,
+                or a quantum circuit composed of linear gates only
+                (currently supported gates are CX and SWAP).
```

Comment:
Using dense arrays doesn't seem super scalable here; given that you've derived it from `Gate`, it's presumably intended to be a permutation matrix, which means that there should only be `n` non-zero entries in the `n * n` matrix.  That's generally far better represented in sparse form - I imagine it'd be better to go that route for scalability?

You don't even need a sparse matrix for this, really - it can be instantiated from a 1D list, since that's a mapping of `index -> new_index`.  Depending on whether you need to be able to invert the function efficiently, the underlying data structure can either be just that list (as a Numpy array), or a pair of lists - one forwards, one backwards.  The unitarity checks and squareness checks become trivial in that case; both are tested by `len(set(new_indices)) == len(new_indices)`.

Reply:
Thanks! As per @ShellyGarion's comment below, the n x n matrix corresponding to a linear function can have an arbitrary number of 0s and 1s. Permutations correspond to very special linear matrices, with a single 1 in every row and column. A somewhat related question is do we want to have some method to detect that a given matrix corresponds to a permutation, and possibly use the much faster algorithm for synthesizing permutations (in `permutation.py`)? And if so, should this be a part of `LinearFunction` or maybe a yet another transpiler pass that changes LinearFunctions to Permutations when possible.

Reply:
Thanks! I have added a new (very simple) transpiler pass to promote linear functions to permutations when possible.